File: tests/utf8_tests.c
#include "../utf8.h"

#include <sys/wait.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static void fail(const char *message) {
  fprintf(stderr, "utf8_tests: %s\n", message);
  exit(1);
}

static void expect_unicode(const char *name, const unsigned char *bytes, size_t len,
                           unicode expected) {
  FILE *file = tmpfile();
  if (file == NULL) {
    fail("tmpfile failed");
  }
  if (fwrite(bytes, 1, len, file) != len) {
    fail("fwrite failed");
  }
  rewind(file);
  unicode actual = fromUTF8(file);
  fclose(file);
  if (actual != expected) {
    fprintf(stderr, "utf8_tests: %s expected %#lx got %#lx\n", name, expected, actual);
    exit(1);
  }
}

static void expect_out(const char *name, unicode value, const unsigned char *expected,
                       size_t expected_len) {
  FILE *file = tmpfile();
  if (file == NULL) {
    fail("tmpfile failed");
  }
  outUTF8(value, file);
  rewind(file);

  unsigned char actual[8];
  size_t actual_len = fread(actual, 1, sizeof(actual), file);
  fclose(file);
  if (actual_len != expected_len || memcmp(actual, expected, expected_len) != 0) {
    fprintf(stderr, "utf8_tests: %s output mismatch\n", name);
    exit(1);
  }
}

static void call_from_utf8(const unsigned char *bytes, size_t len) {
  FILE *file = tmpfile();
  if (file == NULL) {
    fail("tmpfile failed");
  }
  if (fwrite(bytes, 1, len, file) != len) {
    fail("fwrite failed");
  }
  rewind(file);
  (void)fromUTF8(file);
  fclose(file);
}

static void call_out_utf8(unicode value) {
  FILE *file = tmpfile();
  if (file == NULL) {
    fail("tmpfile failed");
  }
  outUTF8(value, file);
  fclose(file);
}

static void expect_fatal(const char *name, void (*call)(const void *), const void *arg,
                         const char *expected_stderr) {
  int stderr_pipe[2];
  if (pipe(stderr_pipe) != 0) {
    fail("pipe failed");
  }

  pid_t pid = fork();
  if (pid < 0) {
    fail("fork failed");
  }
  if (pid == 0) {
    close(stderr_pipe[0]);
    dup2(stderr_pipe[1], STDERR_FILENO);
    close(stderr_pipe[1]);
    call(arg);
    _exit(0);
  }

  close(stderr_pipe[1]);
  char stderr_text[256];
  size_t used = 0;
  for (;;) {
    ssize_t count = read(stderr_pipe[0], stderr_text + used, sizeof(stderr_text) - used - 1);
    if (count < 0) {
      fail("read failed");
    }
    if (count == 0) {
      break;
    }
    used += (size_t)count;
    if (used + 1 == sizeof(stderr_text)) {
      break;
    }
  }
  close(stderr_pipe[0]);
  stderr_text[used] = '\0';

  int status = 0;
  if (waitpid(pid, &status, 0) != pid) {
    fail("waitpid failed");
  }
  if (!WIFEXITED(status) || WEXITSTATUS(status) != 1) {
    fprintf(stderr, "utf8_tests: %s expected exit 1\n", name);
    exit(1);
  }
  if (strcmp(stderr_text, expected_stderr) != 0) {
    fprintf(stderr, "utf8_tests: %s stderr mismatch\nexpected: %sactual: %s", name,
            expected_stderr, stderr_text);
    exit(1);
  }
}

struct bytes_case {
  const unsigned char *bytes;
  size_t len;
};

static void fatal_from_utf8(const void *arg) {
  const struct bytes_case *bytes = arg;
  call_from_utf8(bytes->bytes, bytes->len);
}

static void fatal_out_utf8(const void *arg) {
  const unicode *value = arg;
  call_out_utf8(*value);
}

int main(void) {
  const unsigned char ascii[] = {0x41};
  const unsigned char two_byte[] = {0xc2, 0xa3};
  const unsigned char three_byte[] = {0xe2, 0x82, 0xac};
  const unsigned char four_byte[] = {0xf0, 0x9f, 0x98, 0x80};
  const unsigned char invalid_lead[] = {0xff};
  const unsigned char incomplete_two_byte[] = {0xc2};

  expect_unicode("ascii", ascii, sizeof(ascii), 0x41);
  expect_unicode("two-byte", two_byte, sizeof(two_byte), 0x00a3);
  expect_unicode("three-byte", three_byte, sizeof(three_byte), 0x20ac);
  expect_unicode("four-byte", four_byte, sizeof(four_byte), 0x1f600);

  expect_out("ascii", 0x41, ascii, sizeof(ascii));
  expect_out("two-byte", 0x00a3, two_byte, sizeof(two_byte));
  expect_out("three-byte", 0x20ac, three_byte, sizeof(three_byte));
  expect_out("four-byte", 0x1f600, four_byte, sizeof(four_byte));

  const struct bytes_case invalid_lead_case = {invalid_lead, sizeof(invalid_lead)};
  const struct bytes_case incomplete_two_byte_case = {incomplete_two_byte,
                                                      sizeof(incomplete_two_byte)};
  const unicode out_of_range = UMAX + 1;
  expect_fatal("invalid lead byte", fatal_from_utf8, &invalid_lead_case,
               "protocol error - invalid sequence: 0xff\n");
  expect_fatal("incomplete two-byte sequence", fatal_from_utf8, &incomplete_two_byte_case,
               "protocol error - incomplete sequence: 0xc2 EOF\n");
  expect_fatal("out of range output", fatal_out_utf8, &out_of_range,
               "char 0x110000 out of unicode range\n");

  return 0;
}

File: tests/utf8_tests.zig
const std = @import("std");
const testing = std.testing;

const utf8 = @import("utf8");

var mock_read_buf: []const u8 = &.{};
var mock_read_idx: usize = 0;

var mock_write_buf: [100]u8 = undefined;
var mock_write_len: usize = 0;

export fn getc(fil: ?*utf8.FILE) c_int {
    _ = fil;
    if (mock_read_idx >= mock_read_buf.len) {
        return -1;
    }
    const byte = mock_read_buf[mock_read_idx];
    mock_read_idx += 1;
    return byte;
}

export fn putc(ch: c_int, fil: ?*utf8.FILE) c_int {
    _ = fil;
    if (mock_write_len < mock_write_buf.len) {
        mock_write_buf[mock_write_len] = @intCast(ch);
        mock_write_len += 1;
    }
    return ch;
}

test "fromUTF8 - successful decoding" {
    const Case = struct {
        bytes: []const u8,
        expected: c_ulong,
    };
    const cases = [_]Case{
        .{ .bytes = &.{0x41}, .expected = 0x41 }, // ascii
        .{ .bytes = &.{0xc2, 0xa3}, .expected = 0x00a3 }, // two-byte
        .{ .bytes = &.{0xe2, 0x82, 0xac}, .expected = 0x20ac }, // three-byte
        .{ .bytes = &.{0xf0, 0x9f, 0x98, 0x80}, .expected = 0x1f600 }, // four-byte
    };

    for (cases) |c| {
        mock_read_buf = c.bytes;
        mock_read_idx = 0;
        const actual = utf8.fromUTF8(null);
        try testing.expectEqual(c.expected, actual);
    }
}

test "outUTF8 - successful encoding" {
    const Case = struct {
        val: c_ulong,
        expected: []const u8,
    };
    const cases = [_]Case{
        .{ .val = 0x41, .expected = &.{0x41} },
        .{ .val = 0x00a3, .expected = &.{0xc2, 0xa3} },
        .{ .val = 0x20ac, .expected = &.{0xe2, 0x82, 0xac} },
        .{ .val = 0x1f600, .expected = &.{0xf0, 0x9f, 0x98, 0x80} },
    };

    for (cases) |c| {
        mock_write_len = 0;
        utf8.outUTF8(c.val, null);
        try testing.expectEqualSlices(u8, c.expected, mock_write_buf[0..mock_write_len]);
    }
}

fn fatalFromUTF8(bytes: []const u8) void {
    mock_read_buf = bytes;
    mock_read_idx = 0;
    _ = utf8.fromUTF8(null);
}

fn fatalOutUTF8(val: c_ulong) void {
    utf8.outUTF8(val, null);
}

fn runExpectFatal(action: anytype, arg: anytype, expected_stderr: []const u8) !void {
    var pipe_fds: [2]c_int = undefined;
    _ = std.posix.system.pipe(&pipe_fds);

    const S = struct {
        extern fn fork() c_int;
    };
    const pid = S.fork();
    if (pid == -1) return error.ForkFailed;

    if (pid == 0) {
        // Child
        _ = std.posix.system.close(pipe_fds[0]);
        _ = std.posix.system.dup2(pipe_fds[1], std.posix.STDERR_FILENO);
        _ = std.posix.system.close(pipe_fds[1]);

        action(arg);
        std.process.exit(0);
    } else {
        // Parent
        _ = std.posix.system.close(pipe_fds[1]);
        var buf: [512]u8 = undefined;
        var total_read: usize = 0;
        while (true) {
            const rc = std.posix.system.read(pipe_fds[0], buf[total_read..].ptr, buf[total_read..].len);
            if (rc < 0) return error.ReadFailed;
            if (rc == 0) break;
            total_read += @intCast(rc);
        }
        _ = std.posix.system.close(pipe_fds[0]);

        var status: c_int = 0;
        _ = std.posix.system.waitpid(pid, @ptrCast(&status), 0);
        try testing.expect(status != 0);
        try testing.expectEqualStrings(expected_stderr, buf[0..total_read]);
    }
}


test "fromUTF8/outUTF8 - fatal cases" {
    // invalid lead byte
    try runExpectFatal(fatalFromUTF8, &.{0xff}, "protocol error - invalid sequence: 0xff\n");

    // incomplete two-byte sequence
    try runExpectFatal(fatalFromUTF8, &.{0xc2}, "protocol error - incomplete sequence: 0xc2 EOF\n");

    // out of range output (0x10ffff + 1 = 0x110000)
    try runExpectFatal(fatalOutUTF8, 0x110000, "char 0x110000 out of unicode range\n");
}
